### ISSUE TITLE
[ban-types] Should allow type parameters to extend `object`

### ISSUE THREAD

**Issue by tajakobsen** [typescript-eslint/typescript-eslint#2720]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Problem**

I think **@typescript-eslint/ban-types** triggers an error on what is a valid usecase.

I want to narrow my type parameters on a function with `<A extends object>` (and disallow other primitive types). And when using the function I give it an `interface` as type parameter `A`.

The [stated objective of the rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md#default-options) is:
> Avoid the object type, as it is currently hard to use due to not being able to assert that keys exist. See microsoft/TypeScript#21732.

This isn't the case when we are not trying to use the `object` value (which is unsafe), but only **use it to narrow** a type parameter.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/ban-types": "error"
  }
}
```

```TS
interface Foo {
  bar: number;
}

function f<A extends object>(a: A): void { // HERE! The `object`triggers "@typescript-eslint/ban-types" error
  return undefined;
}

f<Foo>({ bar: 1 });

f<string>("Hi"); // This fails to compile, as it should
```

**Expected Result**

No lint rule is triggered

**Actual Result**

> 14:40  error  Don't use `object` as a type. The `object` type is currently hard to use ([see this issue](https://github.com/microsoft/TypeScript/issues/21732)).
> Consider using `Record<string, unknown>` instead, as it allows you to more easily inspect and use the keys  @typescript-eslint/ban-types

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.6.0` |
| `@typescript-eslint/parser`        | `4.6.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.12.1` |
| `node`                             | `12.19.0` |

**Comment by bradzacher** [comment]
Ban types just bans you from using a type name. It doesn't do anything fancy or have any special branches for specific types. It just reports when you use a configured name.

If you want to use `object`, you can either reconfigure the rule, or use a disable comment.

**Comment by Jamesgt** [comment]
@bradzacher  Adding an option to *where* to ban types is not possible?
We would need to allow it in every type parameter extension, but never in itself.
So global configuration is not an option for this requirement and adding the disable comment might be needed in a lot of cases where we want to enforce a non-primitive type as type parameter.

Also the message tells to use `Record<string, unknown>` but that actually is not equal to `object`, here we would need to use then `{[P in keyof T]: T[P]}` so the original example would become:
```typescript
function f<A extends {[P in keyof A]: A[P]}>(a: A): void {
```
@tajakobsen please check if this fits your needs.

**Comment by tajakobsen** [comment]
@Jamesgt I use the pattern is in [this project](https://github.com/ItemConsulting/enonic-types/blob/main/src/content.ts#L4).

A simplified example of line 4 is this:

```typescript
//  simplified code from project above
export interface ContentLibrary {
  get<A extends object>(params: GetContentParams): Content<A> | null;
}

// Usage
declare const a: ContentLibrary;

const b = a.get<string>({ // this FAILS because string doesn't extend object
  key: ""
});
```

With the suggested change:

```typescript
// changed type parameter to suggestion by @Jamesgt 
export interface ContentLibrary {
  get<A extends {[P in keyof A]: A[P]}>(params: GetContentParams): Content<A> | null;
}

// Usage
declare const a: ContentLibrary;

const b = a.get<string>({ // string doesn't fail anymore, but that was what I wanted
  key: ""
});
```

So if I understood you correctly, this didn't do the trick for me.

**Comment by Jamesgt** [comment]
we used it like this:
```typescript
type ObjectOf<T> = { [P in keyof T]: T[P] };

type Event<T extends ObjectOf<T>> { data: T };
```
and that ensures that `data` will be non-primitive and with the given type.
Hope that it helps.
